// bakong-khqr/src/constant/index.js
const khqrData = require("./khqrData");
const errorCode = require("./errorCode")

const emv = {
    PAYLOAD_FORMAT_INDICATOR: "00",
    DEFAULT_PAYLOAD_FORMAT_INDICATOR: "01",
    POINT_OF_INITIATION_METHOD: "01",
    STATIC_QR: "11",
    DYNAMIC_QR: "12",
    MERCHANT_ACCOUNT_INFORMATION_INDIVIDUAL: "29",
    MERCHANT_ACCOUNT_INFORMATION_MERCHANT: "30",
    BAKONG_ACCOUNT_IDENTIFIER: "00",
    MERCHANT_ACCOUNT_INFORMATION_MERCHANT_ID: "01",
    INDIVIDUAL_ACCOUNT_INFORMATION: "01",
    MERCHANT_ACCOUNT_INFORMATION_ACQUIRING_BANK: "02",
    MERCHANT_CATEGORY_CODE: "52",
    DEFAULT_MERCHANT_CATEGORY_CODE: "5999",
    TRANSACTION_CURRENCY: "53",
    TRANSACTION_AMOUNT: "54",
    DEFAULT_TRANSACTION_AMOUNT: "0",
    COUNTRY_CODE: "58",
    DEFAULT_COUNTRY_CODE: "KH",
    MERCHANT_NAME: "59",
    MERCHANT_CITY: "60",
    DEFAULT_MERCHANT_CITY: "Phnom Penh",
    CRC: "63",
    CRC_LENGTH: "04",
    ADDITIONAL_DATA_TAG: "62",
    BILLNUMBER_TAG: "01",
    ADDITIONAL_DATA_FIELD_MOBILE_NUMBER: "02",
    STORELABEL_TAG: "03",
    TERMINAL_TAG: "07",
    PURPOSE_OF_TRANSACTION: "08",
    TIMESTAMP_TAG: "99",
    MERCHANT_INFORMATION_LANGUAGE_TEMPLATE: "64",
    LANGUAGE_PREFERENCE: "00",
    MERCHANT_NAME_ALTERNATE_LANGUAGE: "01",
    MERCHANT_CITY_ALTERNATE_LANGUAGE: "02",
    UNIONPAY_MERCHANT_ACCOUNT: "15",
    INVALID_LENGTH: {
        KHQR: 12,
        MERCHANT_NAME: 25,
        BAKONG_ACCOUNT: 32,
        AMOUNT: 13,
        COUNTRY_CODE: 3,
        MERCHANT_CATEGORY_CODE: 4,
        MERCHANT_CITY: 15,
        TIMESTAMP: 13,
        TRANSACTION_AMOUNT: 14,
        TRANSACTION_CURRENCY: 3,
        TIMESTAMP: 13,
        BILL_NUMBER: 25,
        STORE_LABEL: 25,
        TERMINAL_LABEL: 25,
        PURPOSE_OF_TRANSACTION: 25,
        MERCHANT_ID: 32,
        ACQUIRING_BANK: 32,
        MOBILE_NUMBER: 25,
        ACCOUNT_INFORMATION: 32,
        MERCHANT_INFORMATION_LANGUAGE_TEMPLATE: 99,
        UPI_MERCHANT: 99,
        LANGUAGE_PREFERENCE: 2,
        MERCHANT_NAME_ALTERNATE_LANGUAGE: 25,
        MERCHANT_CITY_ALTERNATE_LANGUAGE: 15
    },
};

module.exports = { emv, errorCode, khqrData };
